entity-framework/core/testing/index.md

```diff
@@ -1,117 +1,116 @@
-Instead, EF Core is a common set of patterns and concepts that can be used with any database system.
-EF Core database providers then layer database-specific behavior and functionality over this common framework.
-This allows each database system to do what it does best while still maintaining commonality, where appropriate, with other database systems.
+However, unit tests do not guarantee that your program works correctly as a whole; while each component may work well in isolation, various issues may arise when integrating them together. For example, a component may execute some query against a database, expecting certain behavior, while the actual database system behaves differently. Such issues are only uncovered in integration testing, making them an important part of any testing strategy. In addition, since unit tests interact with internal components, they must be reworked as the program is refactored and internal APIs change; integration tests, in contrast, are more stable as they test public, external API contracts.
 
-Fundamentally, this means that switching out the database provider will change EF Core behavior and the application can't be expected to function correctly unless it explicitly accounts for any differences in behavior.
-That being said, in many cases doing this will work because there is a high degree of commonality amongst relational databases.
-This is good and bad.
-Good because moving between database systems can be relatively easy.
-Bad because it can give a false sense of security if the application is not fully tested against the new database system.
+## Integration testing
 
-## Approach 1: Production database system
+Because of the above difficulties with integration tests, developers are frequently urged to concentrate on unit tests first, and have a robust unit test suite which they can run frequently on their machines; integration tests, in contrast, are supposed to be executed much less frequently, and in many cases also provide far less coverage. We recommend giving more thought to integration tests, and that that databases are actually far less affected by the above problems as much as people tend to think:
 
-As described in the previous section, the only way to be sure you are testing what runs in production is to use the same database system.
-For example, if the deployed application uses SQL Azure, then testing should also be done against SQL Azure.
+1. Most databases can nowadays be easily installed on the developer's machine. Container-based technologies such as Docker can make this very easy, and technologies such as [Github Workspaces](https://docs.github.com/en/codespaces/overview) and [Dev Container](https://code.visualstudio.com/docs/remote/create-dev-container) set up your entire development environment for you (including the database). When using SQL Server, it's also possible to test against [LocalDB](/sql/database-engine/configure-windows/sql-server-express-localdb).
```

Comment:
LocalDb is Windows-only.

Reply:
But SQL Server Express is xplat, and available as a docker image

Reply:
Changing last sentence to: "When using SQL Server, it's also possible to test against [LocalDB](/sql/database-engine/configure-windows/sql-server-express-localdb) on Windows, or easily set up a Docker image on Linux."